### ISSUE TITLE
SSL_CTX_set_ciphersuites accepts TLS1.2 and below suite names

### ISSUE THREAD

**Issue by sam-github** [openssl/openssl#8000]
I found this while reading the source, and confirmed with CLI. Unfortunately, the selected non 1.3 ciphers don't seem to actually be used.

```
% ./out/Release/openssl-cli ciphers -stdname -ciphersuites TLS_PSK_WITH_AES_128_CBC_SHA:TLS_AES_256_GCM_SHA384 ":"

TLS_PSK_WITH_AES_128_CBC_SHA - PSK-AES128-CBC-SHA      SSLv3 Kx=PSK      Au=PSK  Enc=AES(128)  Mac=SHA1
TLS_AES_256_GCM_SHA384 - TLS_AES_256_GCM_SHA384  TLSv1.3 Kx=any      Au=any  Enc=AESGCM(256) Mac=AEAD
```
This is for openssl1.1.1a.

Ideally, the suites selected by standard name would be useable if their protocol was negotiated, but if that isn't going to happen, then the code should be changed to work as documented, and only accept TLS1.3 cipher suite names.

**Comment by sam-github** [comment]
No response, maybe this isn't such a high priority bug :-).

However, I suggest that this API be officially changed to accept TLS1.2 cipher suite names (but only when provided with their *standard* names). This would allow "modern" implementations to use just set_ciphersuites(). It might also allow simpler interop with BoringSSL, which apparently has gone its own way with TLS1.3 configuration APIs, see https://github.com/nodejs/node/pull/26365

**Comment by davidben** [comment]
> BoringSSL, which apparently has gone its own way with TLS1.3 configuration API

BoringSSL does not have an API to configure this to begin with. If the need ever comes up, we'll consider adding it, but we do not generally add such knobs unless there's a clear need so we can tailor it to that.

**Comment by kroeckx** [comment]
The reason for the split for TLS 1.3 ciphersuites is because
people think it's a good idea to hard code ciphers, or put them in
config files. And I start to feel like it was a bad idea to give
people the posibility to do so again.

Boringssl does not have this problem, because Google only
supports boringssl for it's own products and can just fix those
products in case they hard coded ciphers. We had to do something
because too many things broke that we can't fix.

**Comment by sam-github** [comment]
This is likely the wrong place to discuss the BoringSSL API, but am I to understand it doesn't allow any configuration, at all, of TLS1.3 cipher suites? Interesting.

I still suggest it would be useful to have set_ciphersuites() accept any standard names, and do the right thing, irrespective of BoringSSL (my suggestion in the original bug report):

FYI, to deal with this API split in Node.js, I split the cipher list on `:`, and create two lists, one with the `TLS_` ciphers (the TLS1.3 ones), and the other with the rest (the TLS1.2 patterns), then send one list into ciphersuites() and one into cipher_list(). This works OK, I guess, but it feels a bit odd, like something that OpenSSL's APIs could have done for me. The docs don't say, but I'm guessing they don't is because any existing apps calling set_cipher_list() would then effectively never do TLS1.3, and further guessing that the intention was that TLS1.2 apps start getting TLS1.3 automatically unless they opt-out in a new TLS1.3 specific way?

**Comment by davidben** [comment]
> Boringssl does not have this problem, because Google only supports boringssl for it's own products and can just fix those products in case they hard coded ciphers.

No, BoringSSL also detaches the TLS 1.3 ciphers from the TLS 1.2 API. Recall, it was me who suggested this implementation strategy to you.

**Comment by davidben** [comment]
> This is likely the wrong place to discuss the BoringSSL API, but am I to understand it doesn't allow any configuration, at all, of TLS1.3 cipher suites? Interesting.

That's right. There's all of three cipher suites in TLS 1.3 and they're all fine. We don't like exposing configuration APIs without a clear need because it complicates changes in the future. (For instance, [CECPQ2](https://www.imperialviolet.org/2018/12/12/cecpq2.html) wants to be correlated with 256-bit ciphers but is a NamedGroup, not a cipher suite. With hard-coded preferences, we have full flexibility to implement whatever negotiation algorithm is appropriate for this.)

> The docs don't say, but I'm guessing they don't is because any existing apps calling set_cipher_list() would then effectively never do TLS1.3, and further guessing that the intention was that TLS1.2 apps start getting TLS1.3 automatically unless they opt-out in a new TLS1.3 specific way?

The latter. Clients which configure TLS 1.2 ciphers won't know to configure TLS 1.3 ciphers, yet the default maximum version switches to TLS 1.3. This means existing applications would break. (TLS versions are negotiated first, then other parameters. This is necessary for the downgrade protection story to work.)

Thus, the existing configuration option is detached from TLS 1.3 altogether. This also makes sense independent of compatibility. TLS 1.3 cipher suites are totally disjoint from TLS 1.2 and are really completely different animals. One specifies a full handshake shape, while the other specifies just the AEAD and KDF hash. They occupy the same code point space and the same ClientHello field, but there was no fundamental reason to do that. We could just as easily have added a supported_aeads extension to TLS 1.3, ignored the existing ClientHello field, and introduced a new code point space.

You may wish to take this thinking into account when deciding how to expose TLS 1.3 in Node, otherwise you may accidentally introduce the same compatibility problem yourself. This sort of mess is why we don't like exposing TLS configuration options, particularly low-level ones that cannot be ported across versions. It unnecessarily complicates the compatibility story. Exposing TLS 1.3 cipher configuration may not be the right thing for Node to do at all.

**Comment by kaduk** [comment]
While it's tempting to use set_ciphersuites as an opportunity to move away from the baroque openssl-specific names to the standard names, given the current documentation and implementation details I think it would be better to just add a check to the ciphersuite_cb that skips ciphers that are not in the TLS 1.3 universe (i.e., specify only bulk AEAD and hash, without key agreement).  The weird semantics where this API's ordering takes priority over (even server-preference) the other APIs has hard to document, perhaps fragile to maintain across future changes, and potentially confusing.

**Comment by sam-github** [comment]
@davidben its not clear to me whether you are describing boring or open ssl, or both above.

To be clear, Node.js does not support BoringSSL. Some projects, like Electron, choose to use it instead of OpenSSL, and we will merge unobtrusive changes they PR to Node.js if we deem the support cost low, but that is the end of the support. I'm not exactly sure how they will work around the lack of set_ciphersuites() in BoringSSL once Node.js has TLS1.3 support and uses that API, but its not a primary concern of mine (because Node.js doesn't support BoringSSL).  I pointed @nornagon here as a heads up.


Back to OpenSSL and Node.js:

Node.js supports selection of ciphersuites as well as max&min protocol version. We choose a default we consider balances security and interoperability, but Node isn't an application, its a runtime, and if our users are working in a constrained env, and must use insecure TLS protocol versions or cipher suites to interoperate with existing infrastructure, they may configure node to do so, at their discretion and risk. Conversely, if they want to restrict suites and protocol versions to an even more secure subset for internal, regulatory, or whatever constraints, we allow them to do so.

The default for node 10.x and 11.x will be min of tls1, max of tls1.2, but users can enable tls1.3 explicitly (globally, or per connection).

The default for node 12.x will be min of tls1.2, max of tls1.3, but users can reset the min as low as tls1, and reset the max to tls1.2 if they need to disable TLS1.3 (again, globally, or per connection). Having TLS1.3 default to enabled is of course a (mildly) breaking change, which is why it won't come out until 12.x, the next semver-major.

I don't see how we could *not* expose TLS1.3 cipher suite configuration. I expect most people will use the defaults (for TLS1.3, that will be the same 3 suites as OpenSSL enables by default), and IoT or other users wanting to enable the other 2 TLS1.3 suites will be able to enable them explicitly.

I described part of how I implemented  this behaviour in https://github.com/openssl/openssl/issues/8000#issuecomment-468503334, as API feedback to OpenSSL to give them a sense of how the new APIs are being used.. The strategy works for us, we don't require any changes to the OpenSSL APIs to implement it.

But like I said, it feels like the kind of thing the set_ciphersuites() API could easily do itself, and I'm not sure why it doesn't.

**Comment by kroeckx** [comment]
Since you're talking about defaults, openssl.cfg can also specify
such defaults. I use that in Debian to set system wide defaults.

I'm just not sure how such defaults should interact. I guess I
prefer that the openssl.cfg settings were taking into account, and
not just overwritten. That is, if I set a minimum of TLSv1.2 in
openssl.cfg, and nodejs does not have anything in the config file
related to it, that TLSv1.2 is the default minimum, but if needed
you can override that to TLSv1 in the nodejs config file. Or we
could have it the other way that the openssl.cfg has TLSv1 in it,
that nodejs wants to default to TLSv1.2 without having it in the
config file, and that nodejs then takes the highest value of both.
Anything in nodejs's config file overrides whatever is in
openssl.cfg. And the application can the override that.

**Comment by davidben** [comment]
@sam-github The second half of my reply describes common behavior to both libraries. This is a very important API decision around TLS 1.3 and ciphers. If you do attempt to wrap OpenSSL's new API and are playing additional games on top of it, you will need to think very carefully about the ramifications. 

If you do it wrong, enabling TLS 1.3 by default becomes not a minor breaking change, but a major one. Specifically, consider if an application configures a TLS 1.2 cipher list, but expresses no opinions on the maximum TLS version. If that does not pull in the usual TLS 1.3 cipher list by default, this application will break, due to the version and cipher negotiation ordering.

(This is why we opted not to add a TLS 1.3 cipher suite configuration option to BoringSSL. Low-level options like this are a very bad idea. Rather, find out why you need to configure them, and add high-level policies that you can evaluate at all TLS versions. If we add a set_ciphersuites function as a compatibility hack, it will almost certainly be a no-op.)

**Comment by sam-github** [comment]
@davidben  Thanks for the feedback, I am trying to understand your concerns, but so far, your specific concerns don't seem to apply.

A TLS1.2 app that explicitly specifies a cipher list that contains no TLS1.3 ciphers will, even after upgrading to a TLS1.3 capable node, have its max TLS version clamped to TLS1.2, and will not break much less change behaviour.

Feel free to comment on the PR if its of interest to you:
- https://github.com/nodejs/node/pull/26209

The diff is still large, it temporarily includes the info callback patch, the patch will go once openssl-1.1.1b lands, the cipher config handling is quite locallized:
- https://github.com/nodejs/node/blob/5bd81cedd74a412b019e65be23936546ba3ff15f/lib/_tls_common.js#L158-L177

**Comment by sam-github** [comment]
@kroeckx for  better or worse, node.js doesn't use any openssl config files. I've only seen one bug report on this (someone wanting to change the sigalg), but mostly this guarantees node.js uservices can be deployed to various machines, and not change behaviour at the whim of some global config file. The other way of looking at it is they can't be deployed and be customized sensibly for the deployment env, but that doesn't seem to be the experience, particularly not with cloud deployments. Like statically linking in OpenSSL and the other node library deps, this is a decision that far predates my participation in node.js, I'm just interpreting what I think the history is for this decision.

**Comment by kroeckx** [comment]
OpenSSL in 1.1.1 changed to read the config file by default, so
that we can read such defaults.

**Comment by sam-github** [comment]
Thanks for the heads up @kroeckx , I'll look into that.

**Comment by mattcaswell** [comment]
Added the "won't fix" label. I don't expect to make any changes as a result of this issue.

**Comment by paulidale** [comment]
Closing this since changes are very unlikely to happen almost a year after.

**Comment by jordanbrown0** [comment]
It is frustrating that SSL_CTX_set_ciphersuites is documented to accept only TLS 1.3 ciphers.  It *does* accept TLS 1.2 ciphers, and they work, and everything fits together almost exactly the way I would expect - but the documentation doesn't allow it.  Sure, misconfiguration leads to failures; the list of ways to misconfigure TLS is a long one and this is only one item on it.  Forcing these into two separate lists means that consumers need duplicate handling of them, extended all the way out to their user interfaces.

Why do we feel the need to let the user configure the list of TLS 1.3 ciphers?  Because today they are "all fine", and tomorrow somebody will publish a paper breaking an algorithm and implementation, and the day after that the calls will start coming in asking to disable the bad algorithm.  Maybe the algorithms and implementations are all perfect and will never need to be disabled, but that's not the way to bet.

What happens when TLS 1.4 comes along and drops all of the existing cipher suites in favor of a new set?  (Post-quantum crypto might do exactly that.)  Will there be a third function?

I would just maintain one list myself and split it at the last moment, but it's logistically difficult to do that last-minute split.

I'm happy to discuss designs for managing compatibility in this area, if there's interest.
